contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,233 @@
+
+* It is making something better, although not a perfect solution.
+* It can mitigate the pain of the user.
+* There is no bug introduced by the change itself.
+
+Then the code can be accepted.
+
+Of course, code review should have some minimal requirements.
+
+For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.
+
+For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.
+
+## What to look for in a code review
+
+### Design
+
+The most important thing to cover in a review is the overall design.
+
+Read the description of the PR, and if it's not easy to understand, ask the author to improve it. You may consider questions like:
+
+* Does the change integrate well with the rest of our system?
+* Is it a good time now to add this functionality?
+
+[TiDB Internals forum](https://internals.tidb.io/) is a place for discussion. For pull requests that are created without prior consensus, try to seek consensus there.
+
+### Functionality
+
+* Does this PR do what the developer intend to do?
+* Is what the developer intend to do good for the users with this implementation?
+
+Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.
+
+You can always validate the PR if you want. It's hard to understand how some changes will impact a user when you're just reading the code, such as a UI change. You can ask the PR author to give you a demo of the functionality if it's too inconvenient to try it yourself.
+
+### Logic
+
+In the case that concurrent programming involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.
+
+The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.
+
+### Complexity
+
+Is the PR more complex than it should be? Check the following items for review the complexity of a PR:
+
+* Are individual lines too complex?
+* Are functions too complex?
+* Are classes too complex?
+
+"Too complex" usually means "can't be understood quickly by readers". It may also mean "developers are likely to introduce bugs when they try to call or modify this code".
+
+A particular type of complexity is over-engineering, where developers have made the code more generic than it needs to be, or added functionality that isn’t presently needed by the system. Reviewers should be especially vigilant about over-engineering. Encourage developers to solve problems must to be solved now, not problems that a developer speculates might need to be solved in the future. 
```

Comment:
... to solve problems that must be solved now -- not problems ...

Reply:
```suggestion
A particular type of complexity is over-engineering, which means that PR authors have made the code more generic than it needs to be, or added functionality that is not presently needed by the system. Reviewers should be especially vigilant about over-engineering. The TiDB community encourages PR authors to solve the problems that must be solved now -- not the problems that might need to be solved in the future. 
```